frontend/src/hooks/usePaginationRange.ts

```diff
@@ -0,0 +1,118 @@
+/**
+ * Custom hook to generate page ranges for any pagination component.
+ * Referenced from
+ * https://www.freecodecamp.org/news/build-a-custom-pagination-component-in-react/
+ */
+
+import { useMemo } from 'react'
+import range from 'lodash/range'
+
+interface UsePaginationRangeProps<T extends unknown = string> {
+  /**
+   * Number of pages to display to left and right of current page.
+   */
+  siblingCount: number
+
+  /**
+   * Total number of elements to paginate.
+   */
+  totalCount: number
+
+  /**
+   * Size of each page. Determines the number of rendered page counts.
+   */
+  pageSize: number
+
+  /**
+   * Represents the current active page. Note that a `1-based index` is used.
+   */
+  currentPage: number
+
+  /**
+   * A constant to denote a separator in the pagination component.
+   */
+  separator: T
+}
+
+export const usePaginationRange = <T extends unknown = string>({
+  totalCount,
+  pageSize,
+  siblingCount,
+  currentPage,
+  separator,
+}: UsePaginationRangeProps<T>): (T | number)[] => {
+  const paginationRange = useMemo(() => {
+    const totalPageCount = Math.ceil(totalCount / pageSize)
+
+    // Pages count is determined as siblingCount + firstPage + lastPage + currentPage + 2*separator
+    const totalPageNumbers = siblingCount + 5
```

Comment:
Agreed on both counts.

I think the siblingCount formula is just accidentally missing `2*`
```suggestion
    const totalPageNumbers = 2 * siblingCount + 5
```

Reply:
This makes me think the storybook should have additional reference cases to show 1 before and 1 after separators are introduced.

With current code, if we had a storybook case for "exactly 7 pages" (with current page = 1). then I think the display would be incorrect like this:

![image](https://user-images.githubusercontent.com/935223/129324090-77124f0d-b86f-4ea3-aa67-14715bbf3a8c.png)

So to make sure we match the design, I think we need cases like these:
* less than 7 (shows 6 pages, no separators
* exactly 7 (shows 7 pages no separators)
* exactly 8 (shows 5 pages, separator, and page 8)

Reply:
Thanks for the catch, will add the stories and update the logic as described.

Reply:
@mantariksh regarding the variable naming; it's not really the total number of page slots too, since it's actually counting the total number of pages to display; i.e 70 items, 10 items per page = 7 pages. 

Maybe I'll continue to call it totalPageCount, but change totalPageNumbers to numPageButtonSlots

Reply:
https://www.youtube.com/watch?v=nQY4dIxY1H4

Reply:
https://martinfowler.com/bliki/TwoHardThings.html